From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp: basic_streambuf copy constructor and assignment, basic_ios::set_rdbuf.

The copy constructor and assignment operator of basic_streambuf (char,
wchar_t and unsigned short; protected since msvcp100, public in msvcp60)
copy the six get/put area pointers and the locale, the copy getting its own
pointer fields as _Init() sets them up (C++ [streambuf.cons],
[streambuf.assign]); the unsigned short copy keeps its own vtable.
basic_ios::set_rdbuf (msvcp100 and later) sets the buffer without clearing
the stream state, unlike rdbuf(sb). The msvcp_win and msvcp120_app entries
forward to msvcp140 and msvcp120. All were spec stubs; Audacity calls the
copy constructor.
---
 dlls/msvcp100/msvcp100.spec         |  54 +++++++-------
 dlls/msvcp110/msvcp110.spec         |  54 +++++++-------
 dlls/msvcp120/msvcp120.spec         |  54 +++++++-------
 dlls/msvcp120_app/msvcp120_app.spec |  54 +++++++-------
 dlls/msvcp140/msvcp140.spec         |  54 +++++++-------
 dlls/msvcp60/ios.c                  |  82 +++++++++++++++++++++
 dlls/msvcp60/msvcp60.spec           |  24 +++---
 dlls/msvcp90/ios.c                  | 143 ++++++++++++++++++++++++++++++++++++
 dlls/msvcp_win/msvcp_win.spec       |  54 +++++++-------
 9 files changed, 399 insertions(+), 174 deletions(-)

diff --git a/dlls/msvcp100/msvcp100.spec b/dlls/msvcp100/msvcp100.spec
index 2d1ae1e..e5702c5 100644
--- a/dlls/msvcp100/msvcp100.spec
+++ b/dlls/msvcp100/msvcp100.spec
@@ -106,27 +106,27 @@
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
-@ stub -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@XZ(ptr) basic_streambuf_char_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ(ptr) basic_streambuf_char_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ(ptr) basic_streambuf_char_ctor
-@ stub -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@XZ(ptr) basic_streambuf_short_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@XZ(ptr) basic_streambuf_short_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@XZ(ptr) basic_streambuf_short_ctor
-@ stub -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
@@ -541,15 +541,15 @@
 @ stub -arch=arm ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAAAAV01@$$QAV01@@Z
 @ stub -arch=i386 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAEAAV01@$$QAV01@@Z
 @ stub -arch=win64 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAAAEAV01@$$QEAV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_assign
 @ cdecl -arch=arm ??4?$numeric_limits@C@std@@QAAAAV01@ABV01@@Z(ptr ptr) std_Num_base_op_assign
 @ thiscall -arch=i386 ??4?$numeric_limits@C@std@@QAEAAV01@ABV01@@Z(ptr ptr) std_Num_base_op_assign
 @ cdecl -arch=win64 ??4?$numeric_limits@C@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) std_Num_base_op_assign
@@ -3378,15 +3378,15 @@
 @ cdecl -arch=win64 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA?AV?$fpos@H@2@V32@H@Z(ptr ptr ptr long) basic_streambuf_wchar_seekpos
 @ cdecl ?set_new_handler@std@@YAP6AXXZH@Z(long) set_new_handler_reset
 @ cdecl ?set_new_handler@std@@YAP6AXXZP6AXXZ@Z(ptr) set_new_handler
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
 @ cdecl ?setbase@std@@YA?AU?$_Smanip@H@1@H@Z(ptr long) setbase
 @ cdecl -arch=arm ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAAPAV12@PAD_J@Z(ptr ptr int64) basic_streambuf_char_setbuf
 @ thiscall -arch=i386 ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAEPAV12@PAD_J@Z(ptr ptr int64) basic_streambuf_char_setbuf
diff --git a/dlls/msvcp110/msvcp110.spec b/dlls/msvcp110/msvcp110.spec
index f3a2195..0ab61cd 100644
--- a/dlls/msvcp110/msvcp110.spec
+++ b/dlls/msvcp110/msvcp110.spec
@@ -118,27 +118,27 @@
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
-@ stub -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@XZ(ptr) basic_streambuf_char_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ(ptr) basic_streambuf_char_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ(ptr) basic_streambuf_char_ctor
-@ stub -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@XZ(ptr) basic_streambuf_short_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@XZ(ptr) basic_streambuf_short_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@XZ(ptr) basic_streambuf_short_ctor
-@ stub -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
@@ -565,15 +565,15 @@
 @ stub -arch=arm ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAAAAV01@$$QAV01@@Z
 @ stub -arch=i386 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAEAAV01@$$QAV01@@Z
 @ stub -arch=win64 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@$$QEAV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_assign
 @ cdecl -arch=arm ??4Init@ios_base@std@@QAAAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ thiscall -arch=i386 ??4Init@ios_base@std@@QAEAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ cdecl -arch=win64 ??4Init@ios_base@std@@QEAAAEAV012@AEBV012@@Z(ptr ptr) ios_base_Init_op_assign
@@ -3289,15 +3289,15 @@
 @ thiscall -arch=i386 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAE?AV?$fpos@H@2@V32@H@Z(ptr ptr long long int64 long long long) basic_streambuf_wchar_seekpos
 @ cdecl -arch=win64 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA?AV?$fpos@H@2@V32@H@Z(ptr ptr ptr long) basic_streambuf_wchar_seekpos
 @ cdecl ?set_new_handler@std@@YAP6AXXZP6AXXZ@Z(ptr) set_new_handler
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
 @ cdecl ?setbase@std@@YA?AU?$_Smanip@H@1@H@Z(ptr long) setbase
 @ cdecl -arch=arm ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAAPAV12@PAD_J@Z(ptr ptr int64) basic_streambuf_char_setbuf
 @ thiscall -arch=i386 ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAEPAV12@PAD_J@Z(ptr ptr int64) basic_streambuf_char_setbuf
diff --git a/dlls/msvcp120/msvcp120.spec b/dlls/msvcp120/msvcp120.spec
index 8863641..6b86d51 100644
--- a/dlls/msvcp120/msvcp120.spec
+++ b/dlls/msvcp120/msvcp120.spec
@@ -118,27 +118,27 @@
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
-@ stub -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@XZ(ptr) basic_streambuf_char_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ(ptr) basic_streambuf_char_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ(ptr) basic_streambuf_char_ctor
-@ stub -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@XZ(ptr) basic_streambuf_short_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@XZ(ptr) basic_streambuf_short_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@XZ(ptr) basic_streambuf_short_ctor
-@ stub -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
@@ -565,15 +565,15 @@
 @ stub -arch=arm ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAAAAV01@$$QAV01@@Z
 @ stub -arch=i386 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAEAAV01@$$QAV01@@Z
 @ stub -arch=win64 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@$$QEAV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_assign
 @ cdecl -arch=arm ??4Init@ios_base@std@@QAAAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ thiscall -arch=i386 ??4Init@ios_base@std@@QAEAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ cdecl -arch=win64 ??4Init@ios_base@std@@QEAAAEAV012@AEBV012@@Z(ptr ptr) ios_base_Init_op_assign
@@ -3230,15 +3230,15 @@
 @ thiscall -arch=i386 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAE?AV?$fpos@H@2@V32@H@Z(ptr ptr long long int64 long long long) basic_streambuf_wchar_seekpos
 @ cdecl -arch=win64 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA?AV?$fpos@H@2@V32@H@Z(ptr ptr ptr long) basic_streambuf_wchar_seekpos
 @ cdecl ?set_new_handler@std@@YAP6AXXZP6AXXZ@Z(ptr) set_new_handler
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
 @ cdecl ?setbase@std@@YA?AU?$_Smanip@H@1@H@Z(ptr long) setbase
 @ cdecl -arch=arm ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAAPAV12@PAD_J@Z(ptr ptr int64) basic_streambuf_char_setbuf
 @ thiscall -arch=i386 ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAEPAV12@PAD_J@Z(ptr ptr int64) basic_streambuf_char_setbuf
diff --git a/dlls/msvcp120_app/msvcp120_app.spec b/dlls/msvcp120_app/msvcp120_app.spec
index 081de4a..77d66d3 100644
--- a/dlls/msvcp120_app/msvcp120_app.spec
+++ b/dlls/msvcp120_app/msvcp120_app.spec
@@ -118,27 +118,27 @@
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@_N@Z(ptr long long long) msvcp120.??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@_N@Z
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@_N@Z(ptr long long long) msvcp120.??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@_N@Z
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z(ptr long long long) msvcp120.??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z
-@ stub -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z(ptr ptr) msvcp120.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z
+@ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z(ptr ptr) msvcp120.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z
+@ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) msvcp120.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) msvcp120.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@W4_Uninitialized@1@@Z
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) msvcp120.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@W4_Uninitialized@1@@Z
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) msvcp120.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@W4_Uninitialized@1@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@XZ(ptr) msvcp120.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@XZ
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ(ptr) msvcp120.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ(ptr) msvcp120.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ
-@ stub -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z(ptr ptr) msvcp120.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z
+@ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z(ptr ptr) msvcp120.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z
+@ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) msvcp120.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) msvcp120.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@W4_Uninitialized@1@@Z
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) msvcp120.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@W4_Uninitialized@1@@Z
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) msvcp120.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@W4_Uninitialized@1@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@XZ(ptr) msvcp120.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@XZ
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@XZ(ptr) msvcp120.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@XZ
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@XZ(ptr) msvcp120.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@XZ
-@ stub -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z(ptr ptr) msvcp120.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z
+@ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z(ptr ptr) msvcp120.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z
+@ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) msvcp120.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) msvcp120.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@W4_Uninitialized@1@@Z
 @ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) msvcp120.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@W4_Uninitialized@1@@Z
 @ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) msvcp120.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@W4_Uninitialized@1@@Z
@@ -565,15 +565,15 @@
 @ stub -arch=arm ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAAAAV01@$$QAV01@@Z
 @ stub -arch=i386 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAEAAV01@$$QAV01@@Z
 @ stub -arch=win64 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@$$QEAV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) msvcp120.??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z
+@ thiscall -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) msvcp120.??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z
+@ cdecl -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) msvcp120.??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) msvcp120.??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z
+@ thiscall -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) msvcp120.??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z
+@ cdecl -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) msvcp120.??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) msvcp120.??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z
+@ thiscall -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) msvcp120.??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z
+@ cdecl -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) msvcp120.??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z
 @ cdecl -arch=arm ??4Init@ios_base@std@@QAAAAV012@ABV012@@Z(ptr ptr) msvcp120.??4Init@ios_base@std@@QAAAAV012@ABV012@@Z
 @ thiscall -arch=i386 ??4Init@ios_base@std@@QAEAAV012@ABV012@@Z(ptr ptr) msvcp120.??4Init@ios_base@std@@QAEAAV012@ABV012@@Z
 @ cdecl -arch=win64 ??4Init@ios_base@std@@QEAAAEAV012@AEBV012@@Z(ptr ptr) msvcp120.??4Init@ios_base@std@@QEAAAEAV012@AEBV012@@Z
@@ -3230,15 +3230,15 @@
 @ thiscall -arch=i386 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAE?AV?$fpos@H@2@V32@H@Z(ptr ptr long long int64 long long long) msvcp120.?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAE?AV?$fpos@H@2@V32@H@Z
 @ cdecl -arch=win64 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA?AV?$fpos@H@2@V32@H@Z(ptr ptr ptr long) msvcp120.?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA?AV?$fpos@H@2@V32@H@Z
 @ cdecl ?set_new_handler@std@@YAP6AXXZP6AXXZ@Z(ptr) msvcp120.?set_new_handler@std@@YAP6AXXZP6AXXZ@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) msvcp120.?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) msvcp120.?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) msvcp120.?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) msvcp120.?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) msvcp120.?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) msvcp120.?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) msvcp120.?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) msvcp120.?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) msvcp120.?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
 @ cdecl ?setbase@std@@YA?AU?$_Smanip@H@1@H@Z(ptr long) msvcp120.?setbase@std@@YA?AU?$_Smanip@H@1@H@Z
 @ cdecl -arch=arm ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAAPAV12@PAD_J@Z(ptr ptr int64) msvcp120.?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAAPAV12@PAD_J@Z
 @ thiscall -arch=i386 ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAEPAV12@PAD_J@Z(ptr ptr int64) msvcp120.?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAEPAV12@PAD_J@Z
diff --git a/dlls/msvcp140/msvcp140.spec b/dlls/msvcp140/msvcp140.spec
index 672f991..beaad80 100644
--- a/dlls/msvcp140/msvcp140.spec
+++ b/dlls/msvcp140/msvcp140.spec
@@ -133,27 +133,27 @@
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
-@ stub -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_char_ctor_uninitialized
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@XZ(ptr) basic_streambuf_char_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ(ptr) basic_streambuf_char_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ(ptr) basic_streambuf_char_ctor
-@ stub -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@XZ(ptr) basic_streambuf_short_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@XZ(ptr) basic_streambuf_short_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@XZ(ptr) basic_streambuf_short_ctor
-@ stub -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_wchar_ctor_uninitialized
@@ -572,15 +572,15 @@
 @ stub -arch=arm ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAAAAV01@$$QAV01@@Z
 @ stub -arch=i386 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAEAAV01@$$QAV01@@Z
 @ stub -arch=win64 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@$$QEAV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_assign
 @ cdecl -arch=arm ??4Init@ios_base@std@@QAAAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ thiscall -arch=i386 ??4Init@ios_base@std@@QAEAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ cdecl -arch=win64 ??4Init@ios_base@std@@QEAAAEAV012@AEBV012@@Z(ptr ptr) ios_base_Init_op_assign
@@ -3197,15 +3197,15 @@
 @ thiscall -arch=i386 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAE?AV?$fpos@U_Mbstatet@@@2@V32@H@Z(ptr ptr long long int64 long long long) basic_streambuf_wchar_seekpos
 @ cdecl -arch=win64 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA?AV?$fpos@U_Mbstatet@@@2@V32@H@Z(ptr ptr ptr long) basic_streambuf_wchar_seekpos
 @ cdecl ?set_new_handler@std@@YAP6AXXZP6AXXZ@Z(ptr) set_new_handler
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) basic_ios_char_set_rdbuf
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) basic_ios_wchar_set_rdbuf
 @ cdecl ?setbase@std@@YA?AU?$_Smanip@H@1@H@Z(ptr long) setbase
 @ cdecl -arch=arm ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAAPAV12@PAD_J@Z(ptr ptr int64) basic_streambuf_char_setbuf
 @ thiscall -arch=i386 ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAEPAV12@PAD_J@Z(ptr ptr int64) basic_streambuf_char_setbuf
diff --git a/dlls/msvcp60/ios.c b/dlls/msvcp60/ios.c
index e54161a..9518d17 100644
--- a/dlls/msvcp60/ios.c
+++ b/dlls/msvcp60/ios.c
@@ -1030,6 +1030,47 @@ char* __thiscall basic_streambuf_char_pptr(const basic_streambuf_char *this)
     return *this->pwpos;
 }
 
+/* The public copy constructor and assignment copy the six buffer pointers and
+ * the locale; the copy gets its own pointer fields, as _Init() sets them up. */
+static void basic_streambuf_char_copy_state(basic_streambuf_char *this, const basic_streambuf_char *copy)
+{
+    basic_streambuf_char_setg(this, basic_streambuf_char_eback(copy),
+            basic_streambuf_char_gptr(copy), basic_streambuf_char_egptr(copy));
+    basic_streambuf_char_setp_next(this, basic_streambuf_char_pbase(copy),
+            basic_streambuf_char_pptr(copy), basic_streambuf_char_epptr(copy));
+}
+
+/* ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE@ABV01@@Z */
+/* ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_char_copy_ctor, 8)
+basic_streambuf_char* __thiscall basic_streambuf_char_copy_ctor(basic_streambuf_char *this,
+        const basic_streambuf_char *copy)
+{
+    TRACE("(%p %p)\n", this, copy);
+
+    this->vtable = &basic_streambuf_char_vtable;
+    locale_copy_ctor(IOS_LOCALE(this), IOS_LOCALE(copy));
+    basic_streambuf_char__Init_empty(this);
+    basic_streambuf_char_copy_state(this, copy);
+    return this;
+}
+
+/* ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAEAAV01@ABV01@@Z */
+/* ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAAAEAV01@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_char_assign, 8)
+basic_streambuf_char* __thiscall basic_streambuf_char_assign(basic_streambuf_char *this,
+        const basic_streambuf_char *assign)
+{
+    TRACE("(%p %p)\n", this, assign);
+
+    if (this != assign)
+    {
+        basic_streambuf_char_copy_state(this, assign);
+        locale_operator_assign(IOS_LOCALE(this), IOS_LOCALE(assign));
+    }
+    return this;
+}
+
 /* ?pubimbue@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE?AVlocale@2@ABV32@@Z */
 /* ?pubimbue@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA?AVlocale@2@AEBV32@@Z */
 DEFINE_THISCALL_WRAPPER(basic_streambuf_char_pubimbue, 12)
@@ -1630,6 +1671,47 @@ wchar_t* __thiscall basic_streambuf_wchar_pptr(const basic_streambuf_wchar *this
     return *this->pwpos;
 }
 
+/* The public copy constructor and assignment copy the six buffer pointers and
+ * the locale; the copy gets its own pointer fields, as _Init() sets them up. */
+static void basic_streambuf_wchar_copy_state(basic_streambuf_wchar *this, const basic_streambuf_wchar *copy)
+{
+    basic_streambuf_wchar_setg(this, basic_streambuf_wchar_eback(copy),
+            basic_streambuf_wchar_gptr(copy), basic_streambuf_wchar_egptr(copy));
+    basic_streambuf_wchar_setp_next(this, basic_streambuf_wchar_pbase(copy),
+            basic_streambuf_wchar_pptr(copy), basic_streambuf_wchar_epptr(copy));
+}
+
+/* ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE@ABV01@@Z */
+/* ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_short_copy_ctor, 8)
+basic_streambuf_wchar* __thiscall basic_streambuf_short_copy_ctor(basic_streambuf_wchar *this,
+        const basic_streambuf_wchar *copy)
+{
+    TRACE("(%p %p)\n", this, copy);
+
+    this->vtable = &basic_streambuf_short_vtable;
+    locale_copy_ctor(IOS_LOCALE(this), IOS_LOCALE(copy));
+    basic_streambuf_wchar__Init_empty(this);
+    basic_streambuf_wchar_copy_state(this, copy);
+    return this;
+}
+
+/* ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAEAAV01@ABV01@@Z */
+/* ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_wchar_assign, 8)
+basic_streambuf_wchar* __thiscall basic_streambuf_wchar_assign(basic_streambuf_wchar *this,
+        const basic_streambuf_wchar *assign)
+{
+    TRACE("(%p %p)\n", this, assign);
+
+    if (this != assign)
+    {
+        basic_streambuf_wchar_copy_state(this, assign);
+        locale_operator_assign(IOS_LOCALE(this), IOS_LOCALE(assign));
+    }
+    return this;
+}
+
 /* ?pubimbue@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AVlocale@2@ABV32@@Z */
 /* ?pubimbue@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA?AVlocale@2@AEBV32@@Z */
 DEFINE_THISCALL_WRAPPER(basic_streambuf_wchar_pubimbue, 12)
diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index fabb959..e96757b 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -704,18 +704,18 @@
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@XZ(ptr) basic_streambuf_char_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ(ptr) basic_streambuf_char_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ(ptr) basic_streambuf_char_ctor
-@ stub -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAA@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE@ABV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA@AEBV01@@Z(ptr ptr) basic_streambuf_char_copy_ctor
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) basic_streambuf_short_ctor_uninitialized
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@XZ(ptr) basic_streambuf_short_ctor
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@XZ(ptr) basic_streambuf_short_ctor
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@XZ(ptr) basic_streambuf_short_ctor
-@ stub -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAA@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE@ABV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
+@ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA@AEBV01@@Z(ptr ptr) basic_streambuf_short_copy_ctor
 @ cdecl -arch=arm ??0?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QAA@ABV01@@Z(ptr ptr) MSVCP_basic_string_char_copy_ctor
 @ thiscall -arch=i386 ??0?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QAE@ABV01@@Z(ptr ptr) MSVCP_basic_string_char_copy_ctor
 @ cdecl -arch=win64 ??0?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAA@AEBV01@@Z(ptr ptr) MSVCP_basic_string_char_copy_ctor
@@ -1472,12 +1472,12 @@
 @ stub -arch=arm ??4?$basic_ostringstream@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QAAAAV01@ABV01@@Z
 @ stub -arch=i386 ??4?$basic_ostringstream@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QAEAAV01@ABV01@@Z
 @ stub -arch=win64 ??4?$basic_ostringstream@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_char_assign
+@ cdecl -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAAAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ thiscall -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAEAAV01@ABV01@@Z(ptr ptr) basic_streambuf_wchar_assign
+@ cdecl -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) basic_streambuf_wchar_assign
 @ cdecl -arch=arm ??4?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QAAAAV01@ABV01@@Z(ptr ptr) MSVCP_basic_string_char_assign
 @ thiscall -arch=i386 ??4?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QAEAAV01@ABV01@@Z(ptr ptr) MSVCP_basic_string_char_assign
 @ cdecl -arch=win64 ??4?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) MSVCP_basic_string_char_assign
diff --git a/dlls/msvcp90/ios.c b/dlls/msvcp90/ios.c
index 80e7694..2fa7dcb 100644
--- a/dlls/msvcp90/ios.c
+++ b/dlls/msvcp90/ios.c
@@ -1627,6 +1627,57 @@ char* __thiscall basic_streambuf_char_pptr(const basic_streambuf_char *this)
     return *this->pwpos;
 }
 
+/* The copy constructor and assignment (public in msvcp60, protected later)
+ * copy the six buffer pointers and the locale (C++ [streambuf.cons],
+ * [streambuf.assign]); the copy gets its own pointer fields, as _Init() sets
+ * them up. */
+static void basic_streambuf_char_copy_state(basic_streambuf_char *this, const basic_streambuf_char *copy)
+{
+#ifndef SG_MUTANT_SB_COPY_NOGET
+    basic_streambuf_char_setg(this, basic_streambuf_char_eback(copy),
+            basic_streambuf_char_gptr(copy), basic_streambuf_char_egptr(copy));
+#endif
+    basic_streambuf_char_setp_next(this, basic_streambuf_char_pbase(copy),
+            basic_streambuf_char_pptr(copy), basic_streambuf_char_epptr(copy));
+}
+
+/* ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z */
+/* ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_char_copy_ctor, 8)
+basic_streambuf_char* __thiscall basic_streambuf_char_copy_ctor(basic_streambuf_char *this,
+        const basic_streambuf_char *copy)
+{
+    TRACE("(%p %p)\n", this, copy);
+
+    this->vtable = &basic_streambuf_char_vtable;
+#if _MSVCP_VER >= 70 && _MSVCP_VER <= 100
+    mutex_ctor(&this->lock);
+#endif
+#if _MSVCP_VER >= 70
+    this->loc = operator_new(sizeof(locale));
+#endif
+    locale_copy_ctor(IOS_LOCALE(this), IOS_LOCALE(copy));
+    basic_streambuf_char__Init_empty(this);
+    basic_streambuf_char_copy_state(this, copy);
+    return this;
+}
+
+/* ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z */
+/* ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_char_assign, 8)
+basic_streambuf_char* __thiscall basic_streambuf_char_assign(basic_streambuf_char *this,
+        const basic_streambuf_char *assign)
+{
+    TRACE("(%p %p)\n", this, assign);
+
+    if (this != assign)
+    {
+        basic_streambuf_char_copy_state(this, assign);
+        locale_operator_assign(IOS_LOCALE(this), IOS_LOCALE(assign));
+    }
+    return this;
+}
+
 /* ?pubimbue@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE?AVlocale@2@ABV32@@Z */
 /* ?pubimbue@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA?AVlocale@2@AEBV32@@Z */
 DEFINE_THISCALL_WRAPPER(basic_streambuf_char_pubimbue, 12)
@@ -2549,6 +2600,66 @@ wchar_t* __thiscall basic_streambuf_wchar_pptr(const basic_streambuf_wchar *this
     return *this->pwpos;
 }
 
+/* see basic_streambuf_char_copy_state */
+static void basic_streambuf_wchar_copy_state(basic_streambuf_wchar *this, const basic_streambuf_wchar *copy)
+{
+#ifndef SG_MUTANT_SB_COPY_NOGET
+    basic_streambuf_wchar_setg(this, basic_streambuf_wchar_eback(copy),
+            basic_streambuf_wchar_gptr(copy), basic_streambuf_wchar_egptr(copy));
+#endif
+    basic_streambuf_wchar_setp_next(this, basic_streambuf_wchar_pbase(copy),
+            basic_streambuf_wchar_pptr(copy), basic_streambuf_wchar_epptr(copy));
+}
+
+/* ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z */
+/* ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_wchar_copy_ctor, 8)
+basic_streambuf_wchar* __thiscall basic_streambuf_wchar_copy_ctor(basic_streambuf_wchar *this,
+        const basic_streambuf_wchar *copy)
+{
+    TRACE("(%p %p)\n", this, copy);
+
+    this->vtable = &basic_streambuf_wchar_vtable;
+#if _MSVCP_VER >= 70 && _MSVCP_VER <= 100
+    mutex_ctor(&this->lock);
+#endif
+#if _MSVCP_VER >= 70
+    this->loc = operator_new(sizeof(locale));
+#endif
+    locale_copy_ctor(IOS_LOCALE(this), IOS_LOCALE(copy));
+    basic_streambuf_wchar__Init_empty(this);
+    basic_streambuf_wchar_copy_state(this, copy);
+    return this;
+}
+
+/* ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z */
+/* ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_short_copy_ctor, 8)
+basic_streambuf_wchar* __thiscall basic_streambuf_short_copy_ctor(basic_streambuf_wchar *this,
+        const basic_streambuf_wchar *copy)
+{
+    TRACE("(%p %p)\n", this, copy);
+    basic_streambuf_wchar_copy_ctor(this, copy);
+    this->vtable = &basic_streambuf_short_vtable;
+    return this;
+}
+
+/* ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z */
+/* ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_wchar_assign, 8)
+basic_streambuf_wchar* __thiscall basic_streambuf_wchar_assign(basic_streambuf_wchar *this,
+        const basic_streambuf_wchar *assign)
+{
+    TRACE("(%p %p)\n", this, assign);
+
+    if (this != assign)
+    {
+        basic_streambuf_wchar_copy_state(this, assign);
+        locale_operator_assign(IOS_LOCALE(this), IOS_LOCALE(assign));
+    }
+    return this;
+}
+
 /* ?pubimbue@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@QAE?AVlocale@2@ABV32@@Z */
 /* ?pubimbue@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@QEAA?AVlocale@2@AEBV32@@Z */
 /* ?pubimbue@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AVlocale@2@ABV32@@Z */
@@ -5923,6 +6034,22 @@ basic_streambuf_char* __thiscall basic_ios_char_rdbuf_set(basic_ios_char *this,
     return ret;
 }
 
+#if _MSVCP_VER >= 100
+/* ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z */
+/* ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z */
+/* Unlike rdbuf(sb), the stream state is left as it is ([basic.ios.members]). */
+DEFINE_THISCALL_WRAPPER(basic_ios_char_set_rdbuf, 8)
+void __thiscall basic_ios_char_set_rdbuf(basic_ios_char *this, basic_streambuf_char *streambuf)
+{
+    TRACE("(%p %p)\n", this, streambuf);
+
+    this->strbuf = streambuf;
+#ifdef SG_MUTANT_SETRDBUF_CLEAR
+    basic_ios_char_clear(this, IOSTATE_goodbit);
+#endif
+}
+#endif
+
 /* ?rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QBEPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@XZ */
 /* ?rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEBAPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@XZ */
 DEFINE_THISCALL_WRAPPER(basic_ios_char_rdbuf_get, 4)
@@ -6206,6 +6333,22 @@ basic_streambuf_wchar* __thiscall basic_ios_wchar_rdbuf_set(basic_ios_wchar *thi
     return ret;
 }
 
+#if _MSVCP_VER >= 100
+/* ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z */
+/* ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z */
+/* Unlike rdbuf(sb), the stream state is left as it is ([basic.ios.members]). */
+DEFINE_THISCALL_WRAPPER(basic_ios_wchar_set_rdbuf, 8)
+void __thiscall basic_ios_wchar_set_rdbuf(basic_ios_wchar *this, basic_streambuf_wchar *streambuf)
+{
+    TRACE("(%p %p)\n", this, streambuf);
+
+    this->strbuf = streambuf;
+#ifdef SG_MUTANT_SETRDBUF_CLEAR
+    basic_ios_wchar_clear(this, IOSTATE_goodbit);
+#endif
+}
+#endif
+
 /* ?rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QBEPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@XZ */
 /* ?rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEBAPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@XZ */
 /* ?rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QBEPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@XZ */
diff --git a/dlls/msvcp_win/msvcp_win.spec b/dlls/msvcp_win/msvcp_win.spec
index f5cb561..af78e56 100644
--- a/dlls/msvcp_win/msvcp_win.spec
+++ b/dlls/msvcp_win/msvcp_win.spec
@@ -133,27 +133,27 @@
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@_N@Z(ptr long long long) msvcp140.??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@_N@Z
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@_N@Z(ptr long long long) msvcp140.??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@_N@Z
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z(ptr long long long) msvcp140.??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z
-@ stub -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z(ptr ptr) msvcp140.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@ABV01@@Z
+@ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z(ptr ptr) msvcp140.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@ABV01@@Z
+@ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) msvcp140.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@AEBV01@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) msvcp140.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@W4_Uninitialized@1@@Z
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) msvcp140.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@W4_Uninitialized@1@@Z
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) msvcp140.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@W4_Uninitialized@1@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@XZ(ptr) msvcp140.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAA@XZ
 @ thiscall -arch=i386 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ(ptr) msvcp140.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ
 @ cdecl -arch=win64 ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ(ptr) msvcp140.??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ
-@ stub -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z(ptr ptr) msvcp140.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@ABV01@@Z
+@ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z(ptr ptr) msvcp140.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@ABV01@@Z
+@ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) msvcp140.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@AEBV01@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) msvcp140.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@W4_Uninitialized@1@@Z
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) msvcp140.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@W4_Uninitialized@1@@Z
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) msvcp140.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@W4_Uninitialized@1@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@XZ(ptr) msvcp140.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAA@XZ
 @ thiscall -arch=i386 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@XZ(ptr) msvcp140.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAE@XZ
 @ cdecl -arch=win64 ??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@XZ(ptr) msvcp140.??0?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAA@XZ
-@ stub -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z
-@ stub -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z
-@ stub -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z
+@ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z(ptr ptr) msvcp140.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@ABV01@@Z
+@ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z(ptr ptr) msvcp140.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@ABV01@@Z
+@ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z(ptr ptr) msvcp140.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@AEBV01@@Z
 @ cdecl -arch=arm ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@W4_Uninitialized@1@@Z(ptr long) msvcp140.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAA@W4_Uninitialized@1@@Z
 @ thiscall -arch=i386 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@W4_Uninitialized@1@@Z(ptr long) msvcp140.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAE@W4_Uninitialized@1@@Z
 @ cdecl -arch=win64 ??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@W4_Uninitialized@1@@Z(ptr long) msvcp140.??0?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAA@W4_Uninitialized@1@@Z
@@ -571,15 +571,15 @@
 @ stub -arch=arm ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAAAAV01@$$QAV01@@Z
 @ stub -arch=i386 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IAEAAV01@$$QAV01@@Z
 @ stub -arch=win64 ??4?$basic_ostream@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@$$QEAV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) msvcp140.??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAAAAV01@ABV01@@Z
+@ thiscall -arch=i386 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) msvcp140.??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAEAAV01@ABV01@@Z
+@ cdecl -arch=win64 ??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) msvcp140.??4?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) msvcp140.??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAAAAV01@ABV01@@Z
+@ thiscall -arch=i386 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) msvcp140.??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IAEAAV01@ABV01@@Z
+@ cdecl -arch=win64 ??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) msvcp140.??4?$basic_streambuf@GU?$char_traits@G@std@@@std@@IEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z(ptr ptr) msvcp140.??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAAAAV01@ABV01@@Z
+@ thiscall -arch=i386 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z(ptr ptr) msvcp140.??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IAEAAV01@ABV01@@Z
+@ cdecl -arch=win64 ??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z(ptr ptr) msvcp140.??4?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@IEAAAEAV01@AEBV01@@Z
 @ cdecl -arch=arm ??4Init@ios_base@std@@QAAAAV012@ABV012@@Z(ptr ptr) msvcp140.??4Init@ios_base@std@@QAAAAV012@ABV012@@Z
 @ thiscall -arch=i386 ??4Init@ios_base@std@@QAEAAV012@ABV012@@Z(ptr ptr) msvcp140.??4Init@ios_base@std@@QAEAAV012@ABV012@@Z
 @ cdecl -arch=win64 ??4Init@ios_base@std@@QEAAAEAV012@AEBV012@@Z(ptr ptr) msvcp140.??4Init@ios_base@std@@QEAAAEAV012@AEBV012@@Z
@@ -3196,15 +3196,15 @@
 @ thiscall -arch=i386 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAE?AV?$fpos@U_Mbstatet@@@2@V32@H@Z(ptr ptr long long int64 long long long) msvcp140.?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAE?AV?$fpos@U_Mbstatet@@@2@V32@H@Z
 @ cdecl -arch=win64 ?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA?AV?$fpos@U_Mbstatet@@@2@V32@H@Z(ptr ptr ptr long) msvcp140.?seekpos@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA?AV?$fpos@U_Mbstatet@@@2@V32@H@Z
 @ cdecl ?set_new_handler@std@@YAP6AXXZP6AXXZ@Z(ptr) msvcp140.?set_new_handler@std@@YAP6AXXZP6AXXZ@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
-@ stub -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
-@ stub -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) msvcp140.?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) msvcp140.?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXPAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z(ptr ptr) msvcp140.?set_rdbuf@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXPEAV?$basic_streambuf@DU?$char_traits@D@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) msvcp140.?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAAXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) msvcp140.?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QAEXPAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z(ptr ptr) msvcp140.?set_rdbuf@?$basic_ios@GU?$char_traits@G@std@@@std@@QEAAXPEAV?$basic_streambuf@GU?$char_traits@G@std@@@2@@Z
+@ cdecl -arch=arm ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) msvcp140.?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAAXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ thiscall -arch=i386 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) msvcp140.?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QAEXPAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
+@ cdecl -arch=win64 ?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z(ptr ptr) msvcp140.?set_rdbuf@?$basic_ios@_WU?$char_traits@_W@std@@@std@@QEAAXPEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@2@@Z
 @ cdecl ?setbase@std@@YA?AU?$_Smanip@H@1@H@Z(ptr long) msvcp140.?setbase@std@@YA?AU?$_Smanip@H@1@H@Z
 @ cdecl -arch=arm ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAAPAV12@PAD_J@Z(ptr ptr int64) msvcp140.?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAAPAV12@PAD_J@Z
 @ thiscall -arch=i386 ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAEPAV12@PAD_J@Z(ptr ptr int64) msvcp140.?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAEPAV12@PAD_J@Z
